### PR TITLE
4934 validation doesn't consider loaded ig immediately

### PR DESCRIPTION
**Pull request by volodymyr-korzh**
Previously, installing or uninstalling an implementation guide does not refresh the validation cache. This problem has been fixed.

### CODE REVIEW

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java [hunk 1]
```diff
@@ -163,6 +166,65 @@ public void testInstallUsCore() {
 		myPackageInstallerSvc.install(spec);
 	}
 
+	@Test
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:169]: suggested test name:
testValidationCache_whenInstallingIG_isRefreshed

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java [hunk 2]
```diff
@@ -163,6 +166,65 @@ public void testInstallUsCore() {
...
+			.setInstallMode(PackageInstallationSpec.InstallModeEnum.STORE_ONLY);
+
+		myPackageInstallerSvc.install(spec);
+
+		// validation after installation should be successful
+		ValidationResult validationResultAfter = validateWithResult(patient);
+		assertTrue(validationResultAfter.isSuccessful());
+	}
+
+	@Test
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:197]: testValidationCache_whenUnInstallingIG_isRefreshed

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java [hunk 3]
```diff
@@ -163,6 +166,65 @@ public void testInstallUsCore() {
 		myPackageInstallerSvc.install(spec);
 	}
 
+	@Test
+	public void install_validateResourceAfterUkCoreIgInstallation_Success() {
+		// crate resource with UKCore profile
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:171]: the above comment is not required.  what ur doing is obvious.  please remove

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java [hunk 4]
```diff
@@ -163,6 +166,65 @@ public void testInstallUsCore() {
...
+		Patient patient = new Patient();
+		patient.setMeta(new Meta().addProfile("https://fhir.nhs.uk/R4/StructureDefinition/UKCore-Patient"));
+
+		// validate resource before loading IG to initiate validation caches
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:175]: same here

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java [hunk 5]
```diff
@@ -163,6 +166,65 @@ public void testInstallUsCore() {
...
+
+		// install UK.Core package
+		PackageInstallationSpec spec = new PackageInstallationSpec()
+			.setName("UK.Core.r4")
+			.setVersion("1.1.0")
+			.setInstallMode(PackageInstallationSpec.InstallModeEnum.STORE_ONLY);
+
+		myPackageInstallerSvc.install(spec);
+
+		// validation after installation should be successful
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:192]: pls remove

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java [hunk 6]
```diff
@@ -163,6 +166,65 @@ public void testInstallUsCore() {
...
+	public void install_validateResourceAfterUkCoreIgInstallation_Success() {
+		// crate resource with UKCore profile
+		Patient patient = new Patient();
+		patient.setMeta(new Meta().addProfile("https://fhir.nhs.uk/R4/StructureDefinition/UKCore-Patient"));
+
+		// validate resource before loading IG to initiate validation caches
+		ValidationResult validationResultBefore = validateWithResult(patient);
+		assertFalse(validationResultBefore.isSuccessful());
+
+		// initiate jpaPackageCache for IG package loading
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:179]: the following will reach out on the internet and download the package. that may take some time.  also, the operation can be denied on the pipeline.  regardless, see testInstallR4PackageWithNoDescription.  JA has already downloaded the package and made it available to use in ur test.  please modify.

**Review by epeartree — COMMENTED** [review]
Please address comments and ping me when done.  i'd like to know when 'validationSupport' caches are being rebuilt.

thank.

good work.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/4934-validation-doesnt-consider-loaded-ig-immedieately.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+---
+type: fix
+issue: 4934
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/4934-validation-doesnt-consider-loaded-ig-immedieately.yaml:3]: Fixed

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java (hunk 2 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:189]: Fixed

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java (hunk 3 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java]: Fixed

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java (hunk 4 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java]: Removed

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java (hunk 6 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java]: Noted, rewrote package installation according to testInstallR4PackageWithNoDescription

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java [hunk 8]
```diff
@@ -163,6 +166,47 @@ public void testInstallUsCore() {
 		myPackageInstallerSvc.install(spec);
 	}
 
+	@Test
+	public void testValidationCache_whenInstallingIG_isRefreshed() {
```
Comment (jmarchionatto) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:170]: Neat tests!

Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/NpmR4Test.java:170]: Thanks ! :)

**Pull request merged**